client/json-rpc/src/async_client/client.rs

```diff
@@ -2,155 +2,64 @@
-    pub fn get_currencies() -> Self {
-        Self::new("get_currencies", json!([]))
-    }
-
-    pub fn get_events(key: &str, start_seq: u64, limit: u64) -> Self {
-        Self::new("get_events", json!([key, start_seq, limit]))
-    }
-
-    pub fn get_transactions(start_seq: u64, limit: u64, include_events: bool) -> Self {
-        Self::new(
-            "get_transactions",
-            json!([start_seq, limit, include_events]),
-        )
-    }
-
-    pub fn get_account_transactions(
-        address: &AccountAddress,
-        start_seq: u64,
-        limit: u64,
-        include_events: bool,
-    ) -> Self {
-        Self::new(
-            "get_account_transactions",
-            json!([address, start_seq, limit, include_events]),
-        )
-    }
-
-    pub fn get_account_transaction(
-        address: &AccountAddress,
-        seq: u64,
-        include_events: bool,
-    ) -> Self {
-        Self::new(
-            "get_account_transaction",
-            json!([address, seq, include_events]),
-        )
-    }
-
-    pub fn get_account(address: &AccountAddress) -> Self {
-        Self::new("get_account", json!([address]))
-    }
-    pub fn get_metadata_by_version(version: u64) -> Self {
-        Self::new("get_metadata", json!([version]))
-    }
-
-    pub fn get_metadata() -> Self {
-        Self::new("get_metadata", json!([]))
-    }
 
-    pub fn to_json(&self, id: usize) -> serde_json::Value {
-        json!({"jsonrpc": "2.0", "id": id, "method": self.method, "params": self.params})
-    }
-}
-
-#[derive(Debug)]
-pub struct Response<R> {
-    pub result: R,
-    pub state: State,
-}
-
-impl<R> std::ops::Deref for Response<R> {
-    type Target = R;
-
-    fn deref(&self) -> &Self::Target {
-        &self.result
-    }
-}
-
-impl<R: for<'de> serde::Deserialize<'de>> TryFrom<JsonRpcResponse> for Response<R> {
-    type Error = Error;
+use serde::de::DeserializeOwned;
 
-    fn try_from(resp: JsonRpcResponse) -> Result<Self, Error> {
-        let state = State::from_response(&resp);
-        match resp.result {
-            Some(ret) => Ok(Self {
-                result: serde_json::from_value::<R>(ret)
-                    .map_err(Error::DeserializeResponseJsonError)?,
-                state,
-            }),
-            None => Err(Error::ResultNotFound(resp)),
-        }
-    }
-}
+use std::{convert::TryInto, sync::Arc, time::Duration};
 
-#[derive(Debug)]
 pub struct Client<R> {
-    pub http_client: reqwest::Client,
-    pub server_url: reqwest::Url,
-    pub last_known_state: std::sync::RwLock<Option<State>>,
+    pub http_client: Arc<dyn HttpClient>,
     pub retry: R,
 }
 
 impl<R: RetryStrategy> Client<R> {
     pub fn from_url<T: reqwest::IntoUrl>(server_url: T, retry: R) -> Result<Self, reqwest::Error> {
+        let reqwest_client = reqwest::ClientBuilder::new()
+            .use_native_tls()
+            .timeout(defaults::HTTP_REQUEST_TIMEOUT)
+            .build()
+            .expect("Unable to build Client.");
```

Comment:
This looks like an impl detail of `SimpleHttpClient` and this layer probably shouldn't know about it. Should we just expose a `SimpleHttpClient::new` method that takes care of this?

Reply:
Updated